### DISCUSSION TITLE
Calling a ruby function on button click

### DISCUSSION THREAD

**Discussion by i-h8-github** [jeremyevans/roda#345] [General]
Hello all,

How can I call a ruby function when a button is clicked without requesting a new page?

**Comment by jeremyevans** [comment]
I'm not sure what you are asking for here, but I'm guessing you want to have a user click a button, and have some code executed on the server based on the button click, but not have navigation go to a new page.  Assuming that is what you want, you would have to have javascript code on the page that overrides the default button behavior to submit the request via `XMLHttpRequest` or `fetch` to a route in your Roda application, and have that Roda route call the Ruby function, then (if necessary), return appropriate data in the response, and have the javascript handle the response.

**Comment by i-h8-github** [comment]
Thank you. That looks like more or less what Unpoly does, is that right? Seems they even have a Roda plugin ready to go, so I will be checking that out.

Thank you for your reply.

By the way, I really Like Roda (and many of your other projects) a lot. Thank you for your generosity.

**Reply by i-h8-github** [reply]
Okay, I don't understand how this works at all.

**Reply by i-h8-github** [reply]
I can't even get non-GET requests to work without Roda::RodaPlugins::RouteCsrf::InvalidToken

**Reply by i-h8-github** [reply]
Maybe Turbo is better. Who knows? Doesn't really matter if I can't post or whatever.

**Reply by jeremyevans** [reply]
Invalid CSRF token would be because you didn't generate or submit the correct token with the request.  By default, the route_csrf plugin requires request-specific tokens for higher security, but you can turn that off if you want to.

As for Unpoly, I'm not sure as I've never used it.

**Reply by i-h8-github** [reply]
Thank you, just as I was getting at the root of the problem, some kind soul on irc helped me realize that I needed to pass a matching path to the `csrf_tag` method.

I have everything working as expected now. Thank you for your attention and patience.

P.S. Is there a more elegant equivalent of `r.redirect(r.referrer)`? Also, I tried `r.referrer` on a whim after looking at some Rails code, and was pleased to see it work, but I couldn't find where it was documented.

P.S. Do you have an opinion on Turbo?

**Reply by jeremyevans** [reply]
`r.referrer` comes from `Rack::Request`. I don't think there is a more elegant equivalent to what you want. I would be careful when using `r.referrer`, as the information comes from the client, and malicious clients can submit whatever they want in the `Referer` header.

No opinion on Turbo, I haven't used that either.

**Reply by i-h8-github** [reply]
So what would be the recommended way to reroute back to the route the current request came from?

**Reply by jeremyevans** [reply]
To do so securely, you could submit the current path/referrer as a parameter, along with an HMAC parameter using an application secret (possibly the csrf secret).  Here's how you could create the HMAC:

```ruby
r.path # referrer parameter value
OpenSSL::HMAC.digest(OpenSSL::Digest::SHA256.new, csrf_secret, r.path) # HMAC parameter value
```

When the form is submitted, recalculate the HMAC using `r.referrer` instead of `r.path`, and compare it to the submitted HMAC parameter value.  If the HMACs match, you know the referrer parameter hasn't been tampered with by the client, and it is safe to use.

**Reply by i-h8-github** [reply]
Thank you.

I understand the concept, but not the execution. When I try to submit the HMAC as above as a value of a hidden input I get `incompatible character encodings: UTF-8 and ASCII-8BIT`.

**Reply by jeremyevans** [reply]
Try using `HMAC.hexdigest` instead of `HMAC.digest`, that should fix the encoding issue.

**Reply by i-h8-github** [reply]
Thank you again.

With the inclusion of the HMAC, what is the point of sending the path as a parameter?

**Reply by jeremyevans** [reply]
`Referer` header is URL, not path.

**Reply by i-h8-github** [reply]
But if I calculate the HMAC with `r.path` in the form, then again with `r.referrer.sub(r.base_url, "")` in the route, I already have everything I need to determine whether the request is safe, right?

**Reply by jeremyevans** [reply]
`Referer` also includes query string parameters.  I suppose you could strip those out as well, or change the HMAC to include them (you can use `r.fullpath` instead of `r.path` for that).  Note that `r.referrer` may be `nil`. Some browsers are configured to not send `Referer` headers.  I definitely recommend the separate parameter instead of attempting to rely on `Referer` values.

**Reply by i-h8-github** [reply]
Now I'm confused.

In my view, I have:

```
<form method="POST" action="/test">
  <%== csrf_tag("/test", "post") %>
  <input type="hidden" name="key" value="<%= OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, csrf_secret, r.path) %>" />
  <button type="submit">Test</button>
</form>
```

And in my route I have:

```
r.post do
  r.on "test" do
    referrer = r.referrer.sub(r.base_url, "")
    key = OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, csrf_secret, referrer)
    if r.params["key"] == key
      r.redirect referrer
    end
  end
end
```

What am I missing?

**Reply by jeremyevans** [reply]
If your approach works for your use case, feel free to use it.   However, it wouldn't correctly handle a browser that does not send `Referer` headers, and it wouldn't correctly handle the case where the page with the button was requested with query parameters that change the page.

**Reply by i-h8-github** [reply]
How would I accomplish this in a browser that doesn't send referrer headers?

**Reply by jeremyevans** [reply]
You would use a separate form parameter for the path to redirect to.

**Reply by i-h8-github** [reply]
Oh, duh. Doesn't that also render the whole HMAC thing moot? At least if the only point is to know where to go back to after the post and both routes are within my domain?

**Reply by jeremyevans** [reply]
No, you still need the HMAC, because the form parameter can be modified by a malicious client (just as the `Referer` header can be).  The main advantage of the form parameter over the `Referer` header is that all browsers will submit all form parameters, and you have more control over the parameter value.

**Reply by i-h8-github** [reply]
Then wouldn't I recalculate in the `post` block with `r.params["path"]` instead of `r.referrer`?

i.e.

```
<form method="POST" action="/test">
  <%== csrf_tag("/test", "post") %>
  <input type="hidden" name="path" value="<%= r.path %>" />
  <input type="hidden" name="key" value="<%= OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, csrf_secret, r.path) %>" />
  <button type="submit">Test</button>
</form>
```

and

```
r.post do
  r.on "test" do
    key = OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA256.new, csrf_secret, r.params["path"])
    if r.params["key"] == key
      r.redirect r.params["path"]
    end
  end
end
```

**Reply by jeremyevans** [reply]
Yes.

**Reply by i-h8-github** [reply]
Thank you.

You have been extraordinarily patient and helpful.
